### DISCUSSION TITLE
FieldOutOfBoundSurfaceError in OceanParcels

### DISCUSSION THREAD

**Discussion by cristinarumi** [Parcels-code/Parcels#1528] [Q&A]
Hi Erik,

during my simulation I have included the kernel DeleteErrorParticle at the end of the Kernels but I still encountered the error: "File "media/disk/Cristina/ParcelsRun/1993_2018/parcels/parcels/kernel.py", line 587, in execute
raise FieldOutOfBoundSurfaceError (p.lon,p.lat,p.depth)
parcels.tools.statuscodes.<img width="411" alt="Screenshot 2024-04-01 at 09 20 19" src="https://github.com/OceanParcels/parcels/assets/137803399/64f8110a-f1d8-4853-8c6e-e323160dde9d">r: Field sampled out-of-bound at the surface, at (12.091083428090098,n-6.190445023616803, -0.04458211335611706)"

My code is the following:

def DeleteErrorParticle(particle, fieldset, time): #Per les particules que surten de la meva regio
if particle.state == StatusCode.ErrorOutOfBounds:
particle.delete()

def SampleT(particle, fieldset, time):
particle.T = fieldset.T[time, particle.depth, particle.lat, particle.lon]

def SampleS(particle, fieldset, time):
particle.S = fieldset.S[time, particle.depth, particle.lat, particle.lon]

def KeepInOcean(particle, fieldset, time): #per les particules que s'escapen per la superficie
if particle.state == StatusCode.ErrorThroughSurface:
particle_ddepth = 0.52
particle.state = StatusCode.Success

class sParticle(JITParticle):
VT = Variable('VT', initial=0, dtype=np.float32, to_write='once')
T = Variable( 'T', dtype=np.float32)
S = Variable( 'S', dtype=np.float32)

pset = ParticleSet(fieldset=fieldset, pclass=sParticle, lon=lons.astype(float), lat=lats.astype(float), depth=deps.astype(float), time=times, VT=VTS.astype(float))

adv = pset.Kernel(AdvectionRK4_3D)
sampT = pset.Kernel(SampleT)
sampS = pset.Kernel(SampleS)
submerge = pset.Kernel(KeepInOcean)
deleteparticle = pset.Kernel(DeleteErrorParticle)

kernels=adv+submerge+sampT+sampS+deleteparticle

pset.execute(kernels, runtime=delta(days=9491), dt=-delta(days=timest_calc), output_file=ParticleFile(outdir, pset, outputdt=delta(days=timest_out)), verbose_progress=True)

How can I solve the problem? Thank you very much in advanced.

**Comment by erikvansebille** [comment]
Hi @cristinarumi, your error in the screenshot seems to point to a `StatusCode.ErrorThroughSurface` (field out of bounds _at the surface_), but your `DeleteErrorParticle` kernel only captures `StatusCode.ErrorOutOfBounds`. 

Capturing particles that go through the surface requires a different StatusCode ([see this part of the tutorial](https://docs.oceanparcels.org/en/latest/examples/tutorial_kernelloop.html#Working-with-Status-Codes)), so you could try changing your kernel to

```python
def DeleteErrorParticle(particle, fieldset, time):
    if (particle.state == StatusCode.ErrorOutOfBounds) or (particle.state == StatusCode.ErrorThroughSurface):
        particle.delete()
```

**Comment by cristinarumi** [comment]
Hi Erik,
Thank you!
Should I algo move kernel submerge to the end before deleteparticle?




Obtén l'aplicació Outlook per a l'iOS<https://aka.ms/o0ukef>
________________________________
De: Erik van Sebille ***@***.***>
Enviat el: Tuesday, April 2, 2024 8:36:28 AM
Per a: OceanParcels/parcels ***@***.***>
A/c: Cristina Arumí Planas ***@***.***>; Mention ***@***.***>
Tema: Re: [OceanParcels/parcels] FieldOutOfBoundSurfaceError in OceanParcels (Discussion #1528)


Hi @cristinarumi<https://urldefense.com/v3/__https://github.com/cristinarumi__;!!D9dNQwwGXtA!V3ofCnT5445vuXEkBqUsnPUCvQ6pMmewiWRgVqSp7Wks53I2j3rbx5PreOE2VYXj-DhhKpC25eGYtgeChhGLUE7r0UkW$>, your error in the screenshot seems to point to a StatusCode.ErrorThroughSurface (field out of bounds at the surface), but your DeleteErrorParticle kernel only captures StatusCode.ErrorOutOfBounds.

Capturing particles that go through the surface requires a different StatusCode (see this part of the tutorial<https://urldefense.com/v3/__https://docs.oceanparcels.org/en/latest/examples/tutorial_kernelloop.html*Working-with-Status-Codes__;Iw!!D9dNQwwGXtA!V3ofCnT5445vuXEkBqUsnPUCvQ6pMmewiWRgVqSp7Wks53I2j3rbx5PreOE2VYXj-DhhKpC25eGYtgeChhGLUKpWXhuX$>), so you could try changing your kernel to

def DeleteErrorParticle(particle, fieldset, time):
    if (particle.state == StatusCode.ErrorOutOfBounds) or (particle.state == StatusCode.ErrorThroughSurface):
        particle.delete()

**Comment by erikvansebille** [comment]
Well, you only have a kernel `KeepInOcean`, but that should be _after_ `sampT` and `sampS`. All the kernels that deal with errors should normally be at the end of the kernel list

**Reply by cristinarumi** [reply]
Alright, I'll implement these changes. Thanks for your time.
